packages/browser/src/integrations/graphqlClient.ts

```diff
@@ -0,0 +1,107 @@
+  SEMANTIC_ATTRIBUTE_URL_FULL,
+  defineIntegration,
+  spanToJSON,
+} from '@sentry/core';
+import type { Client, IntegrationFn } from '@sentry/types';
+import { parseGraphQLQuery } from '@sentry/utils';
+
+interface GraphQLClientOptions {
+  endpoints: Array<string>;
+}
+
+interface GraphQLRequestPayload {
+  query: string;
+  operationName?: string;
+  variables?: Record<string, any>;
+}
+
+const INTEGRATION_NAME = 'GraphQLClient';
+
+const _graphqlClientIntegration = ((options: GraphQLClientOptions) => {
+  return {
+    name: INTEGRATION_NAME,
+    setup(client) {
+      _updateSpanWithGraphQLData(client, options);
+      _updateBreadcrumbWithGraphQLData(client, options);
+    },
+  };
+}) satisfies IntegrationFn;
+
+function _updateSpanWithGraphQLData(client: Client, options: GraphQLClientOptions): void {
+  client.on('outgoingRequestSpanStart', (span, { body }) => {
+    const spanJSON = spanToJSON(span);
+
+    const spanAttributes = spanJSON.data || {};
+    const spanOp = spanAttributes[SEMANTIC_ATTRIBUTE_SENTRY_OP];
+
+    const isHttpClientSpan = spanOp === 'http.client';
+
+    if (isHttpClientSpan) {
+      const httpUrl = spanAttributes[SEMANTIC_ATTRIBUTE_URL_FULL] || spanAttributes['http.url'];
+
+      const { endpoints } = options;
+      const isTracedGraphqlEndpoint = endpoints.includes(httpUrl);
+
+      if (isTracedGraphqlEndpoint) {
+        const httpMethod = spanAttributes[SEMANTIC_ATTRIBUTE_HTTP_REQUEST_METHOD] || spanAttributes['http.method'];
+
+        const operationInfo = _getGraphQLOperation(body);
+        span.updateName(`${httpMethod} ${httpUrl} (${operationInfo})`);
+        span.setAttribute('body', JSON.stringify(body));
+      }
+    }
+  });
+}
+
+function _updateBreadcrumbWithGraphQLData(client: Client, options: GraphQLClientOptions): void {
+  client.on('outgoingRequestBreadcrumbStart', (breadcrumb, { body }) => {
+    const { category, type, data } = breadcrumb;
+
+    const isFetch = category === 'fetch';
+    const isXhr = category === 'xhr';
+    const isHttpBreadcrumb = type === 'http';
+
+    if (isHttpBreadcrumb && (isFetch || isXhr)) {
+      const httpUrl = data && data.url;
+      const { endpoints } = options;
+
+      const isTracedGraphqlEndpoint = endpoints.includes(httpUrl);
+
+      if (isTracedGraphqlEndpoint && data) {
+        if (!data.graphql) {
+          const operationInfo = _getGraphQLOperation(body);
+
+          data.graphql = {
+            query: (body as GraphQLRequestPayload).query,
+            operationName: operationInfo,
+          };
+        }
+
+        // The body prop attached to HandlerDataFetch for the span should be removed.
+        if (isFetch && data.body) {
+          delete data.body;
+        }
+      }
+    }
+  });
+}
+
+function _getGraphQLOperation(requestBody: unknown): string {
+  // Standard graphql request shape: https://graphql.org/learn/serving-over-http/#post-request
+  const graphqlBody = requestBody as GraphQLRequestPayload;
```

Comment:
if we do not know this is a body, we should at least try-catch all of this. But if we (as mentioned in comments above) use `getBodyString`, we can simply pass this in types as string, as we know it was converted before (or is undefined, in which case, nothing to do)